### PR TITLE
fix: typeof function

### LINKED ISSUE #1181: Typeof function unexpected
While trying to generate schema from property that have a field ConnectionOptions from node:tls the generator fail

Pull request to follow

There is also no FunctionTypeFormatter, added it to avoid failing

I used to test and do the PR this interface

```
import { ConnectionOptions } from "node:tls";

export interface Test extends ConnectionOptions {}

```

### PR DESCRIPTION
**Pull request by loopingz**
Fix issue #1181

### CODE REVIEW

File: test/valid-data/type-typeof-function/main.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { ConnectionOptions } from "node:tls";
+
+export interface Test extends ConnectionOptions {}
```
Comment (domoritz) [inline test/valid-data/type-typeof-function/main.ts]: Please define a smaller test case.

File: src/TypeFormatter/FunctionTypeFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            properties: {
+                isFunction: {
+                    type: "boolean",
+                    const: true,
+                },
+            },
+        };
+    }
+
+    // If this type does NOT HAVE children, generally all you need is:
```
Comment (domoritz) [inline src/TypeFormatter/FunctionTypeFormatter.ts]: What's this comment for?

**Review by domoritz — CHANGES_REQUESTED** [review]
I think we had requests for supporting functions before. Please search the issues and prs. 

in general, I'm not sure I want to support functions since json schema doesn't support them. Maybe the cleaner way is to just ignore functions entirely?

File: src/TypeFormatter/FunctionTypeFormatter.ts (hunk 2 above)
Comment (loopingz) [reply src/TypeFormatter/FunctionTypeFormatter.ts]: This was from one documentation, i can remove it

File: test/valid-data/type-typeof-function/main.ts (hunk 1 above)
Comment (loopingz) [reply test/valid-data/type-typeof-function/main.ts]: I liked the fact that it is really life example

**Comment by loopingz** [comment]
You are right I think they should just be ignored, but failing is not a good option

**Comment by loopingz** [comment]
I did update to return an `UnknownType` with a $comment with the definition of the function.
I think it allows to explain why this property exists while waiting for one day maybe have a normalized way of expressing it with json-schema
If you agree with this approach, I can add it for a FunctionDeclarationNodeParser that would be more generic behavior

**Comment by domoritz** [comment]
Remind me, is $comment standard? Should we add not: {} to indicate that the property should not be set?

**Comment by loopingz** [comment]
The $comment seems a standard to JSONSchema7: http://json-schema.org/draft-07/schema properties.$comment

```
...
    "properties": {
        "$id": {
            "type": "string",
            "format": "uri-reference"
        },
        "$schema": {
            "type": "string",
            "format": "uri"
        },
        "$ref": {
            "type": "string",
            "format": "uri-reference"
        },
        "$comment": {
            "type": "string"
        },
....

```

That was the first behavior to add a `not: {}`, I moved to unknown with `$comment`, I m ok to go back to `not`, we could also add a not with the reasoning within the `$comment`

**Comment by domoritz** [comment]
I think I would prefer no comment in the schema and instead a warning. JSON schema does not support functions and therefore we should not attempt to make them work.

**Comment by loopingz** [comment]
I think adding a `not: {}` is already opiniated, maybe we could add an option to set how to manage function type:

 - `IGNORE`: remove completely the property
 - `COMMENT`: free declaration with a comment on the method signature
 - `FORBID`: add the `not {}` 

With the available override capabilities, we can either just document how to code each behavior

**Comment by domoritz** [comment]
I like the idea of documenting how to implement the behaviors. How about we implement ignore as the default?

**Comment by loopingz** [comment]
Sounds good to me

**Comment by broofa** [comment]
The 'Unknown type "function"' error is going to be pretty common.  This is just one of those differences between TS and JSON that is inherent to this problem space.  But the current behavior is pretty atrocious.  Schema generation fails completely, and you get a really terse error message that has zero information about where the error occurred (The type name and property that triggered the problem is not supplied), and zero information about how to resolve the problem.

Developers are forced to sift through the issues for a solution.  This takes you down the rabbit hole of having to write a custom generator, which is also not particularly well documented.  (This is where I'm at right now, trying to figure out what exactly my generator code needs to look like and having some issues with that, unfortunately.)

I would _much_ rather there were a simple way of indicating how unknown types should be handled as suggested in this issue.  E.g. A config option along the lines of...
```jsonc
  "unknownTypes": {
    "function": "ignore",
    // ...
  }
```

**Comment by kingyue737** [comment]
I still faced the following error in v1.1.2:
```
Error: Unknown type "function"
```

Do I need to add any config to ignore function type?

---------
**EDITED:**
This regression has been confirmed https://github.com/vega/ts-json-schema-generator/issues/1456

**Comment by ChromeQ** [comment]
Confirming this issue is still present in v1.2.0

**Pull request merged**